### PR TITLE
[CLOUD-2555] EAP CD 13 OpenShift Image

### PR DESCRIPTION
**Pull request by luck3y**
https://issues.jboss.org/browse/CLOUD-2555
￼https://issues.jboss.org/browse/CLOUD-2458
https://issues.jboss.org/browse/CLOUD-2531
https://issues.jboss.org/browse/CLOUD-2655
https://issues.jboss.org/browse/CLOUD-2652
https://issues.jboss.org/browse/CLOUD-2690

Signed-off-by: Ken Wills <kwills@redhat.com>

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: dev-overrides.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+modules:
+      repositories:
+          - name: cct_module
+            git:
+                  url: https://github.com/luck3y/cct_module.git
```
Comment (bstansberry) [inline dev-overrides.yaml]: luck3y?

File: dev-overrides.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+                  ref: cd13-dev
+          - name: jboss-eap-modules
+            git:
+                  url: https://github.com/luck3y/jboss-eap-modules.git
```
Comment (bstansberry) [inline dev-overrides.yaml]: luck3y?

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: dev-overrides.yaml (hunk 1 above)
Comment (luck3y) [reply dev-overrides.yaml]: Yeah, I went back and forth on these, on the one hand, it requires --overrides=dev-overrides.yaml when building, OTOH, not defaulting them to the working branch makes it irrelevant if it points to master on both.  I'll update to take those out.

File: docs/templates/eap-cd-amq-persistent-s2i.adoc [hunk 3]
```diff
@@ -243,18 +244,19 @@ for more information.
...
 |`AMQ_USER` | User name for standard broker user. It is required for connecting to the broker. If left empty, it will be generated. | `${MQ_USERNAME}`
 |`AMQ_PASSWORD` | Password for standard broker user. It is required for connecting to the broker. If left empty, it will be generated. | `${MQ_PASSWORD}`
+|`AMQ_ROLE` | AMQ Role for authenticated user | `${MQ_ROLE}`
 |`AMQ_TRANSPORTS` | -- | `${MQ_PROTOCOL}`
-|`AMQ_QUEUES` | Queue names, separated by commas. These queues will be automatically created when the broker starts. Also, they will be made accessible as JNDI resources in EAP. | `${MQ_QUEUES}`
-|`AMQ_TOPICS` | Topic names, separated by commas. These topics will be automatically created when the broker starts. Also, they will be made accessible as JNDI resources in EAP. | `${MQ_TOPICS}`
+|`AMQ_QUEUES` | Queue names, separated by commas. These queues will be automatically created when the broker starts. Also, they will be made accessible as JNDI resources in EAP. Note that all queues used by the application *must* be specified here in order to be created automatically on the remote AMQ broker. | `${MQ_QUEUES}`
+|`AMQ_ADDRESSES` | -- | `${MQ_TOPICS}`
```
Comment (tremes) [inline docs/templates/eap-cd-amq-persistent-s2i.adoc:253]: AMQ_ADDRESSES ? What is it? Shouldn't be AMQ_TOPICS?

File: docs/templates/eap-cd-amq-persistent-s2i.adoc [hunk 4]
```diff
@@ -243,18 +244,19 @@ for more information.
...
 |`MQ_SERIALIZABLE_PACKAGES` | List of packages that are allowed to be serialized for use in ObjectMessage, separated by commas. If your app doesn't use ObjectMessages, leave this blank. This is a security enforcement. For the rationale, see http://activemq.apache.org/objectmessage.html | `${MQ_SERIALIZABLE_PACKAGES}`
 |`AMQ_SPLIT` | Split the data directory for each node in a mesh. | `${AMQ_SPLIT}`
 |`AMQ_MESH_DISCOVERY_TYPE` | The discovery agent type to use for discovering mesh endpoints.  'dns' will use OpenShift's DNS service to resolve endpoints.  'kube' will use Kubernetes REST API to resolve service endpoints.  If using 'kube' the service account for the pod must have the 'view' role, which can be added via 'oc policy add-role-to-user view system:serviceaccount:<namespace>:default' where <namespace> is the project namespace. | `${AMQ_MESH_DISCOVERY_TYPE}`
 |`AMQ_MESH_SERVICE_NAME` | -- | `${APPLICATION_NAME}-amq-mesh`
 |`AMQ_MESH_SERVICE_NAMESPACE` | -- | --
-|`AMQ_STORAGE_USAGE_LIMIT` | The A-MQ storage usage limit | `${AMQ_STORAGE_USAGE_LIMIT}`
+|`AMQ_STORAGE_USAGE_LIMIT` | The AMQ storage usage limit | `${AMQ_STORAGE_USAGE_LIMIT}`
```
Comment (tremes) [inline docs/templates/eap-cd-amq-persistent-s2i.adoc:259]: I am not really sure but I think this still should be `A-MQ` according the naming in eap71 templates. Maybe this has changed for AMQ7???

File: docs/templates/eap-cd-amq-s2i.adoc [hunk 5]
```diff
@@ -218,13 +218,13 @@ for more information.
 |Deployment |Variable name |Description |Example value
 
 .23+| `${APPLICATION_NAME}`
-|`MQ_SERVICE_PREFIX_MAPPING` | -- | `${APPLICATION_NAME}-amq=MQ`
-|`MQ_JNDI` | JNDI name for connection factory used by applications to connect to the broker, e.g. java:/ConnectionFactory | `${MQ_JNDI}`
+|`MQ_SERVICE_PREFIX_MAPPING` | -- | `${APPLICATION_NAME}-amq7=MQ`
+|`MQ_JNDI` | JNDI name for connection factory used by applications to connect to the broker, e.g. java:/DefaultConnectionFactory | `${MQ_JNDI}`
```
Comment (tremes) [inline docs/templates/eap-cd-amq-s2i.adoc]: Shouldn't this be `java:jboss/DefaultJMSConnectionFactory` ?

File: docs/templates/eap-cd-amq-s2i.adoc [hunk 6]
```diff
@@ -25,13 +25,13 @@ https://docs.openshift.org/latest/architecture/core_concepts/templates.html#para
 
 |`APPLICATION_NAME` | -- | The name for the application. | eap-app | True
 |`HOSTNAME_HTTPS` | -- | Custom hostname for https service route.  Leave blank for default hostname, e.g.: secure-<application-name>-<project>.<default-domain-suffix> | -- | False
-|`SOURCE_REPOSITORY_URL` | -- | Git source URI for application | https://github.com/jboss-openshift/openshift-quickstarts.git | True
-|`SOURCE_REPOSITORY_REF` | -- | Git branch/tag reference | 1.3 | False
+|`SOURCE_REPOSITORY_URL` | -- | Git source URI for application | https://github.com/jboss-developer/jboss-eap-quickstarts.git | True
+|`SOURCE_REPOSITORY_REF` | -- | Git branch/tag reference | openshift | False
 |`CONTEXT_DIR` | -- | Path within Git project to build; empty for root project directory. | helloworld-mdb | False
-|`MQ_JNDI` | `MQ_JNDI` | JNDI name for connection factory used by applications to connect to the broker, e.g. java:/ConnectionFactory | java:/ConnectionFactory | False
+|`MQ_JNDI` | `MQ_JNDI` | JNDI name for connection factory used by applications to connect to the broker, e.g. java:/DefaultConnectionFactory | java:/DefaultConnectionFactory | False
```
Comment (tremes) [inline docs/templates/eap-cd-amq-s2i.adoc]: `java:jboss/DefaultJMSConnectionFactory` ?

File: templates/eap-cd-amq-s2i.json [hunk 7]
```diff
@@ -842,7 +842,7 @@
                                         "value": "${MQ_QUEUES}"
                                     },
                                     {
-                                        "name": "AMQ_TOPICS",
+                                        "name": "AMQ_ADDRESSES",
```
Comment (tremes) [inline templates/eap-cd-amq-s2i.json:845]: I might be wrong but I think this still should be AMQ_TOPICS.

File: docs/templates/eap-cd-amq-persistent-s2i.adoc (hunk 3 above)
Comment (luck3y) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:253]: This is a new AMQ 7 thing, addresses abstract both queues and topics. I've tried to make the templates combine both queues and topics into addresses where relevant, but I'll recheck this one.

File: docs/templates/eap-cd-amq-persistent-s2i.adoc (hunk 4 above)
Comment (luck3y) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:259]: The AMQ templates, and the rest of the CE stuff has been changing A-MQ to AMQ, which is the product name. I'll check in the morning what this should be.

File: docs/templates/eap-cd-amq-s2i.adoc (hunk 5 above)
Comment (luck3y) [reply docs/templates/eap-cd-amq-s2i.adoc]: Yes :). I have a final patch that actually uses this in the messaging script.

File: templates/eap-cd-amq-s2i.json (hunk 7 above)
Comment (luck3y) [reply templates/eap-cd-amq-s2i.json:845]: AMQ_TOPICS doesn't exist in the example AMQ template that I used for reference.

https://github.com/jboss-container-images/jboss-amq-7-broker-openshift-image/blob/amq71-dev/templates/amq-broker-71-basic.yaml#L191

**Comment by tremes** [comment]
https://issues.jboss.org/browse/CLOUD-2655 this just came to my mind....just FYI

File: docs/templates/eap-cd-amq-persistent-s2i.adoc (hunk 4 above)
Comment (luck3y) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:259]: See here for the AMQ name: https://github.com/jboss-container-images/jboss-amq-7-broker-openshift-image/blob/amq71-dev/templates/amq-broker-71-basic.yaml#L9

Comment (tremes) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:259]: OK thanks. Then it's OK

File: templates/eap-cd-amq-persistent-s2i.json [hunk 8]
```diff
@@ -65,9 +65,9 @@
         },
         {
             "displayName": "JMS Connection Factory JNDI Name",
-            "description": "JNDI name for connection factory used by applications to connect to the broker, e.g. java:/DefaultConnectionFactory",
+            "description": "JNDI name for connection factory used by applications to connect to the broker, e.g. java:jboss/DefaultJmsConnectionFactory",
             "name": "MQ_JNDI",
-            "value": "java:/DefaultConnectionFactory",
+            "value": "java:jboss/DefaultJmsConnectionFactory",
```
Comment (tremes) [inline templates/eap-cd-amq-persistent-s2i.json]: @luck3y it is `java:jboss/DefaultJMSConnectionFactory`

Comment (luck3y) [reply templates/eap-cd-amq-persistent-s2i.json]: Sigh, sorry @tremes, thats what I get for trying to use sed before I've had enough coffee!

File: docs/templates/eap-cd-amq-persistent-s2i.adoc (hunk 3 above)
Comment (luck3y) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:253]: @tremes See here: https://github.com/jboss-container-images/jboss-amq-7-broker-openshift-image/blob/amq71-dev/modules/amq-launch/added/launch.sh#L30

File: templates/eap-cd-amq-s2i.json (hunk 7 above)
Comment (luck3y) [reply templates/eap-cd-amq-s2i.json:845]: Also, see here: https://github.com/jboss-container-images/jboss-amq-7-broker-openshift-image/blob/amq71-dev/modules/amq-launch/added/launch.sh#L30

And the example for Artemis: https://github.com/apache/activemq-artemis/blob/master/examples/features/standard/topic/src/main/resources/activemq/server0/broker.xml#L51

**Review by bstansberry — APPROVED** [review]
Approved, but as discussed you're checking with Tom Jenkinson re: jta-crash-rec to make sure there's no too-subtle-for-me-to-detect issue from the AMQ 7 change.

**Comment by luck3y** [comment]
Yep, comments here: https://issues.jboss.org/browse/CLOUD-2442

File: docs/templates/eap-cd-amq-persistent-s2i.adoc (hunk 3 above)
Comment (tremes) [reply docs/templates/eap-cd-amq-persistent-s2i.adoc:253]: OK

File: templates/eap-cd-tx-recovery-s2i.json [hunk 9]
```diff
@@ -32,44 +32,122 @@
...
+            "name": "MQ_PASSWORD",
+            "from": "[a-zA-Z0-9]{8}",
+            "generate": "expression",
+            "required": false
+        },
+        {
+            "displayName": "AMQ Role",
+            "description": "AMQ Role for authenticated user",
+            "name": "MQ_ROLE",
+            "value": "admin"
```
Comment (tremes) [inline templates/eap-cd-tx-recovery-s2i.json:136]: So is it required or not? I would say it's not.

File: templates/eap-cd-tx-recovery-s2i.json [hunk 10]
```diff
@@ -32,44 +32,122 @@
...
+        },
+        {
+            "displayName": "JMS Connection Factory JNDI Name",
+            "description": "JNDI name for connection factory used by applications to connect to the broker, e.g. java:jboss/DefaultJMSConnectionFactory",
+            "name": "MQ_JNDI",
+            "value": "java:jboss/DefaultJMSConnectionFactory",
+            "required": false
+        },
+        {
+            "displayName": "Split Data?",
```
Comment (tremes) [inline templates/eap-cd-tx-recovery-s2i.json]: Maybe we should stick with original `Split the data directory?` which seems to me more obvious

File: templates/eap-cd-tx-recovery-s2i.json (hunk 9 above)
Comment (luck3y) [reply templates/eap-cd-tx-recovery-s2i.json:136]: @tremes Talking with the AMQ 7 broker folks, this is something they may implement in the future. I think allowing it to be specified is fine for now. It's not *required*, but will give us flexibility in the configuration connecting to the external broker.

File: templates/eap-cd-tx-recovery-s2i.json (hunk 10 above)
Comment (luck3y) [reply templates/eap-cd-tx-recovery-s2i.json]: I'll double check this, I didn't invent this name, it probably came from the AMQ template. (But having them the same is a good idea).

**Comment by luck3y** [comment]
@tremes FYI this is the template update for jta-crash-recovery. The migration pod was missing the configuration for the remote AMQ.

https://issues.jboss.org/browse/CLOUD-2442

File: templates/eap-cd-tx-recovery-s2i.json (hunk 9 above)
Comment (luck3y) [reply templates/eap-cd-tx-recovery-s2i.json:136]: Also, I should note, it is present in the AMQ71 broker templates.

File: templates/eap-cd-tx-recovery-s2i.json (hunk 10 above)
Comment (luck3y) [reply templates/eap-cd-tx-recovery-s2i.json]: @tremes I've made this usage consistent now, as, 'Split the data directory?'

**Review by luck3y — COMMENTED** [review]
n/a

File: image.yaml [hunk 11]
```diff
@@ -30,11 +30,11 @@ modules:
           - name: cct_module
             git:
                   url: https://github.com/jboss-openshift/cct_module.git
-                  ref: sprint-18
+                  ref: master
```
Comment (luck3y) [inline image.yaml:33]: Due to https://github.com/jboss-container-images/jboss-eap-7-openshift-image/pull/99 coming in late, after I realized we needed to fix the buildability of the CD 12 tag.

**Comment by luck3y** [comment]
@bstansberry @tremes I think all of the corresponding Jira issues are in qe-approved state now, OK with me merging this?

**Comment by bstansberry** [comment]
@luck3y 👍

File: docs/templates/eap-cd-tx-recovery-s2i.adoc [hunk 12]
```diff
@@ -169,6 +194,15 @@ for more information.
 |jolokia | 8778 | `TCP`
 |http | 8080 | `TCP`
 |ping | 8888 | `TCP`
+.8+| `${APPLICATION_NAME}-amq`
+|jolokia | 8778 | `TCP`
```
Comment (tremes) [inline docs/templates/eap-cd-tx-recovery-s2i.adoc]: During this week testing I discovered that this has changed. Jolokia port for AMQ 7 is now 8161 AFAIK. See https://github.com/jboss-container-images/jboss-amq-7-broker-openshift-image/blob/amq71-dev/templates/amq-broker-71-basic.yaml#L27

File: templates/eap-cd-tx-recovery-s2i.json [hunk 13]
```diff
@@ -451,6 +611,185 @@
...
+                                            "/bin/bash",
+                                            "-c",
+                                            "/opt/amq/bin/readinessProbe.sh"
+                                        ]
+                                    }
+                                },
+                                "ports": [
+                                    {
+                                        "name": "jolokia",
+                                        "containerPort": 8778,
```
Comment (tremes) [inline templates/eap-cd-tx-recovery-s2i.json]: Same here

**Comment by tremes** [comment]
@luck3y  Just found one more issue. See my latests comments. Otherwise it's likely OK.

File: docs/templates/eap-cd-tx-recovery-s2i.adoc (hunk 12 above)
Comment (luck3y) [reply docs/templates/eap-cd-tx-recovery-s2i.adoc]: Nice catch, I'll fix and update the build shortly.

File: templates/eap-cd-amq-s2i.json (hunk 7 above)
Comment (rcernich) [reply templates/eap-cd-amq-s2i.json:845]: Has anybody mentioned to them that they've changed the API for the image from 6.x to 7.x?

**Review by rcernich — APPROVED** [review]

File: templates/eap-cd-amq-s2i.json (hunk 7 above)
Comment (luck3y) [reply templates/eap-cd-amq-s2i.json:845]: @rcernich I asked some questions in IRC a while back, but no I haven't specifically asked this.

@vbusch Any idea on this change (AMQ_TOPICS changing to AMQ_ADDRESSES) ? Are there plans to preserve compatibility?

**Review by rcernich — APPROVED** [review]
lgtm

**Comment by rcernich** [comment]
I think this is good to merge.

File: templates/eap-cd-amq-s2i.json (hunk 7 above)
Comment (vbusch) [reply templates/eap-cd-amq-s2i.json:845]: @luck3y, AMQ Broker 7 does not distinguish between JMS and non-JMS configuration.  The thought was that the model differences between AMQ 6 and AMQ Broker 7, should be considered during migration.  https://access.redhat.com/documentation/en-us/red_hat_amq/7.1/html-single/migrating_to_red_hat_amq_7/index#message_address_changes_in_amq_7
@RoddieKieley would know if there is talk about changing it.

Comment (luck3y) [reply templates/eap-cd-amq-s2i.json:845]: @vbusch It works fine, so no complaints from me :)

File: templates/eap-cd-amq-persistent-s2i.json [hunk 14]
```diff
@@ -807,7 +807,7 @@
                                 },
                                 "ports": [
                                     {
-                                        "name": "jolokia",
+                                        "name": "console-jolokia",
```
Comment (luck3y) [inline templates/eap-cd-amq-persistent-s2i.json]: @maschmid @tremes FYI, updated the AMQ broker jolokia to match what the released AMQ templates have.

**Pull request merged**